### ISSUE TITLE
setup.sh not able to config the system

### ISSUE THREAD

**Issue by xinglin** [spdk/spdk#221]
root@atg-s-watney:/home/xingl/spdk/scripts# ./setup.sh
basename: missing operand
Try 'basename --help' for more information.

git commit@23782bc43cc30115569b270ddb7dbf334da813e8

'basename' at the highlighted line seems to cause the problem. 

```
function get_nvme_name_from_bdf {
        set +e
        nvme_devs=`lsblk -d --output NAME | grep "^nvme"`
        set -e
        for dev in $nvme_devs; do
                echo "dev=$dev"
>         bdf=$(basename $(readlink /sys/block/$dev/device/device))
                ...
```

root@atg-s-watney:/home/xingl/spdk/scripts# ls -l /sys/block/nvme0n1/device/device
-r--r--r-- 1 root root 4096 Nov 29 23:31 /sys/block/nvme0n1/device/device
root@atg-s-watney:/home/xingl/spdk/scripts# readlink /sys/block/nvme0n1/device/device
root@atg-s-watney:/home/xingl/spdk/scripts#

readlink command above does not return anything. And then basename command complains about missing operand. Any comments?

-Xing

**Comment by xinglin** [comment]
setup.sh from 17.07.1 is working. So frustrated dealing with this error with the later versions...

root@atg-s-watney:/home/xingl# diff -uNrp spdk-17.07.1/scripts/setup.sh spdk-17.10.1/scripts/setup.sh
--- spdk-17.07.1/scripts/setup.sh	2017-09-13 13:17:01.000000000 -0700
+++ spdk-17.10.1/scripts/setup.sh	2017-11-27 15:59:54.000000000 -0800
@@ -4,11 +4,15 @@ set -e

 rootdir=$(readlink -f $(dirname $0))/..

-function linux_iter_pci {
+function linux_iter_pci_class_code {
 	# Argument is the class code
-	# TODO: More specifically match against only class codes in the grep
-	# step.
-	lspci -mm -n -D | grep $1 | tr -d '"' | awk -F " " '{print $1}'
+	lspci -mm -n -D | awk -v cc="\"$1\"" -F " " '{if (cc ~ $2) print $1}' | tr -d '"'
+}
+
+function linux_iter_pci_dev_id {
+	# Argument 1 is the vendor id
+	# Argument 2 is the device id
+	lspci -mm -n -D | awk -v ven="\"$1\"" -v dev="\"$2\"" -F " " '{if (ven ~ $3 && dev ~ $4) print $1}' | tr -d '"'
 }

 function linux_bind_driver() {
@@ -41,10 +45,38 @@ function linux_bind_driver() {
 	fi
 }

+function linux_unbind_driver() {
+	bdf="$1"
+	ven_dev_id=$(lspci -n -s $bdf | cut -d' ' -f3 | sed 's/:/ /')
+
+	if ! [ -e "/sys/bus/pci/devices/$bdf/driver" ]; then
+		return 0
+	fi
+
+	old_driver_name=$(basename $(readlink /sys/bus/pci/devices/$bdf/driver))
+
+	echo "$ven_dev_id" > "/sys/bus/pci/devices/$bdf/driver/remove_id" 2> /dev/null || true
+	echo "$bdf" > "/sys/bus/pci/devices/$bdf/driver/unbind"
+	echo "$bdf ($ven_dev_id): $old_driver_name -> no driver"
+}
+
 function linux_hugetlbfs_mount() {
 	mount | grep ' type hugetlbfs ' | awk '{ print $3 }'
 }

+function get_nvme_name_from_bdf {
+	set +e
+	nvme_devs=`lsblk -d --output NAME | grep "^nvme"`
+	set -e
+	for dev in $nvme_devs; do
+		bdf=$(basename $(readlink /sys/block/$dev/device/device))
+		if [ "$bdf" = "$1" ]; then
+			eval "$2=$dev"
+			return
+		fi
+	done
+}
+
...

**Comment by jimharris** [comment]
Hi Xing,

What is your Linux distribution and kernel version?

-Jim

**Comment by xinglin** [comment]
Hi Jim, 

Here is the information about my server. 

# cat /etc/issue
Ubuntu 14.04.5 LTS \n \l

# uname -r
3.13.0-24-generic

-Xing

**Comment by jimharris** [comment]
Thanks Xing.  We are going to try to reproduce this here at Intel and then look at how to resolve.  We will keep you posted.  Apologies for the inconvenience.

**Comment by xinglin** [comment]
Hi Seth, 

I confirm that the patch fixes the problem. Thanks for working on addressing this. Hope this will be merged into the main branch soon. 

-Xing

**Comment by dverkamp-intel** [comment]
This has been merged as commit afec8f1476159bb7ecb85ad460909b01e9f295f2 - please let us know if you have any further difficulties.
